### PR TITLE
ENH: Adding Stability Margin with Mach dependency

### PR DESCRIPTION
**Pull request by Gui-FernandesBR**
## Pull request type

Please check the type of change your PR introduces:

- [x] Code base additions (bugfix, features)

## Pull request checklist

Please check if your PR fulfills the following requirements, depending on the type of PR:

- Code base additions (for bug fixes / features):

  - [x] Tests for the changes have been added
  - [x] Docs have been reviewed and added / updated if needed
  - [x] Lint (`black rocketpy`) has passed locally and any fixes were made
  - [x] All tests (`pytest --runslow`) have passed locally

## What is the current behavior?
- The `flight.staticMargin` doesn't seem good enough nowadays. It do not consider the variation of the center of pressure position as the mach number varies during the flight.
- Users usually have troubles to compare stability across different rocket simulators.

## What is the new behavior?
- Implementation followed the instructions/requirements established here: https://github.com/RocketPy-Team/RocketPy/issues/350
- The Rocket and Flight classes now communicates very well to compute more robust `stabilityMargin` attribute, while the trajectory simulation is not being affected.
- Everything thanks to the most magical class of rocketpy: Function

## Does this introduce a breaking change?

- [x] Yes (static margin is no longer a synonym of stability margin in rocketpy).

## Other information
- I tried to do a chart like [this](https://github.com/RocketPy-Team/RocketPy-Hackathon-2022/pull/99#issuecomment-1229254045), but I got no success at all. This can be done in the future.
- Please, review the changes first, and then we update the notebooks.

### CODE REVIEW

File: rocketpy/Rocket.py
Comment (Gui-FernandesBR) [inline rocketpy/Rocket.py:228]: Notice this was duplicated with line 248 (same file)

File: rocketpy/Rocket.py
Comment (Gui-FernandesBR) [inline rocketpy/Rocket.py:403]: as `None` was being returned, this documentation is currently wrong at the default branch 

(sadly)

File: rocketpy/Rocket.py
Comment (phmbressan) [inline rocketpy/Rocket.py:244]: The side comment length is causing the linter to divide this simple line into multiple.

Just a aesthetical change, an equivalent approach would be using `Function(0)`, I believe it makes it clearer that is only a temporary value.

File: rocketpy/plots/rocket_plots.py
Comment (phmbressan) [inline rocketpy/plots/rocket_plots.py:82]: Great suggestion!

**Review by phmbressan — COMMENTED** [review]
Overall, the Mach dependency is a great addition.

This PR still needs to be updated to follow `snake_case` conventions before approval.

**Comment by Gui-FernandesBR** [comment]
I'm waiting for the #395 PR to be merged before solving the conflicts here.

**Comment by MateusStano** [comment]
Hey! I changed a few things, mostly structure wise.

- I re-added static margin because it is still useful (it is just a special case of the stability margin)
- I made stability margin a 2D Function, so now it makes a 3D plot in `Rocket.all_info()`
- Flight calculates the stability margin through the flight and plots it

Think this is ready for a review

**Comment by Gui-FernandesBR** [comment]
Good work, @MateusStano , thank you so much for improving the stability margin calculation in RocketPy, this was a huge gap since the beginning.

I made some small adjusts to improve plots and prints. You can check them by my last 2 commits. 
Some tests that I think you would like to see:
image

image


I'm also sending you a copy of a notebook running the simulation in case you don't wanna run the notebook in your computer.

I want to approve this PR by I was the one who created.
Well, please consider this as resolved/approved, and feel free to merge into develop, there're no breaking changes.

File: rocketpy/rocket/rocket.py
Comment (phmbressan) [inline rocketpy/rocket/rocket.py:515]: I have no problems with this, but out of curiosity why define a `lambda` source and discretize right after?

Wouldn't it work using `Function` operations directly? (same comment to the `stability_margin` evaluation)

**Review by phmbressan — APPROVED** [review]

File: rocketpy/rocket/rocket.py
Comment (MateusStano) [reply rocketpy/rocket/rocket.py:471]: It shouldn't be necessary, but something goes very wrong when adding the `EmptyMotor` in the rocket initialization. We should probably look into it some other time

File: rocketpy/rocket/rocket.py
Comment (MateusStano) [reply rocketpy/rocket/rocket.py:515]: For some reason, static_margin used to break pretty badly with some liquid and hybrid motors. We added this discretization when developing the liquid and hybrid motors, but it might have been done before we started discretizing things in those classes

I just tested it and it seems it can be remover without issues

**Pull request merged**
